### PR TITLE
Implemented pdf building from source

### PR DESCRIPTION
**Pull request by nirmalsinghania2008**
Requires Rst2pdf PDF extension for sphinx.

### CODE REVIEW

File: conf.py [hunk 1]
```diff
@@ -29,7 +29,7 @@
 
 # Add any Sphinx extension module names here, as strings. They can be extensions
 # coming with Sphinx (named 'sphinx.ext.*') or your custom ones.
-extensions = ['sphinx.ext.todo',]
+extensions = ['sphinx.ext.todo','rst2pdf.pdfbuilder']
```
Comment (jaylett) [inline conf.py]: Is this installed by default with sphinx, or is it an additional requirement? If the latter, probably warrants noting in the README.

**Comment by jaylett** [comment]
Do you understand how readthedocs manages to build PDFs for our manual without this? Because I don't, and it'd be good to have some idea in case this causes problems…

Also, can you please reformat your commit message according to the guidance in [this article](https://robots.thoughtbot.com/5-useful-tips-for-a-better-commit-message)? (It's linked from our developer guide as well.) Main things are: first line should be <= 50 characters and tense ('Implement' not 'Implemented'). If you'd done this, the PR would have taken the message automatically from the commit, and you wouldn't have had to edit it. (You can just amend the commit message and then force push to this branch.)

File: Makefile [hunk 2]
```diff
@@ -92,3 +92,8 @@ doctest:
 	$(SPHINXBUILD) -b doctest $(ALLSPHINXOPTS) $(BUILDDIR)/doctest
 	@echo "Testing of doctests in the sources finished, look at the " \
 	      "results in $(BUILDDIR)/doctest/output.txt."
+
+pdf:
+	$(SPHINXBUILD) -b pdf $(ALLSPHINXOPTS) _build/pdf
+	@echo
+	@echo "Build finished. The PDF Version of Documentation is in _build/pdf."
```
Comment (ojwb) [inline Makefile]: There's no newline character on the last line of this file in your new version (as you can see in the github rendering of the patch - hover your mouse cursor over the slashed-zero).

It's best to have text files end with a newline character - some tools don't handle such files nicely.  And in general, invisible differences in files aren't helpful - trailing whitespace on lines is another example.

File: conf.py (hunk 1 above)
Comment (nirmalsinghania2008) [inline conf.py]: It is a sphinx extension and has to be installed additionally.
Will add,how to get it to work in README

File: Makefile (hunk 2 above)
Comment (nirmalsinghania2008) [inline Makefile]: Sphinx uses pdfLaTeX to generate PDF documentation, and Read the Docs uses TeX Live in its build infrastructure. PDF output with rst2pdf is by far not as good as when using LaTeX but it has the real advantage of not requiring you to install a full LaTeX stack on your machine.

The package to install for Latex based pdf generation is "texlive-full"

**Comment by jaylett** [comment]
@nirmalsinghania2008 is it feasible to have a make target to come as close as possible to RTD's output, providing you have the full tools? It may be too much hassle.

Other than that (minor) question, I think it's just adding something to `README` and then squashing and rewording the commits.

**Comment by nirmalsinghania2008** [comment]
to make target close to RTD's, we might be increasing the dependencies for sphinx as texlive-full is quite big package and building pdf using LaTeX is a hassle.
RTD is somewhat automating the whole process
https://jimmyg.org/blog/2009/sphinx-pdf-generation-with-latex.html
also rst2pdf's output is not way behind LaTeX

**Comment by ojwb** [comment]
That blog post is 7 years old though - a lot can change in that time.  It also makes things more complicated than necessary by using a virtualenv, rather than just installing everything from packages.

Having to have most of texlive installed isn't a huge deal - you only need it if you want to make your own PDF of the guide after all.

So I'd say go for whatever gives the better output (and it being consistent with RTD seems a bonus).

FWIW, I have tried making a PDF from the LaTeX in the not too distant past and the main issue I hit was finding myself in the tedious "having to run LaTeX repeatedly until it works" situation, which I think is to do with index references, and I've never really understood how to avoid.

**Comment by nirmalsinghania2008** [comment]
for me:i like the shiny output of rst2pdf more than by pdflatex.
in pdflatex,there is a two step process of converting documentation in .tex and then converting it to .pdf.
plus we need full tex-live package for just building a pdf which can be a hassle for many.
If you want i can work on automating two step process of pdf building using latex.
but i don't see the need much.

File: README.md [hunk 3]
```diff
@@ -15,6 +15,14 @@ documentation.  Confusingly there are several open source projects called
 the one you want.  We aren't certain exactly what version is needed - it
 seems Sphinx 0.6.4 is too old but 1.1.3 is new enough.
 
+For building a pdf version of documentation,you will need to install rst2pdf
+
+https://github.com/rst2pdf/rst2pdf
```
Comment (jaylett) [inline README.md]: We should recommend installing from pip; installing python packages from a git master is frowned upon for various reasons. Also worth pointing out to people that `rsf2pdf` doesn't support python3 at the moment.

**Comment by nirmalsinghania2008** [comment]
@jaylett please check!

**Comment by nirmalsinghania2008** [comment]
@jaylett please check the changes!

**Comment by nirmalsinghania2008** [comment]
@ojwb can you tell your sphinx version using
 `sphinx-build --version`

in the latest versions of sphinx(1.3.x) there is broken support of rst2pdf.
sphinx changed some of it's internals but rst2pdf hasn't updated it accordingly yet.
[https://github.com/sphinx-doc/sphinx/issues/1844](url)

**Comment by ojwb** [comment]
```
Sphinx (sphinx-build) 1.3.6
```

> can we recommend users to use older version of sphinx?

We could, but it's inconvenient to have to install a special older version of something rather than being able to use the version packaged by your distro.  And once you do, you have software installed in a way which doesn't automatically get updated, even with security fixes.

This is the ticket for fixing this in rst2pdf, which has been open close to 5 months with no response at all from the rst2pdf maintainers:

https://github.com/rst2pdf/rst2pdf/issues/556

So we've no idea when (or even if) this is likely to be fixed, which makes me unkeen to rely on this software.

Building via LaTeX seems more appealing.  The dependencies are quite big, but can all just be installed from packages, and as I noted above, you only need it to build a PDF yourself - most people would be happy to just download a prebuilt PDF.

Not sure how typical I am, but FWIW I already have LaTeX installed, but I had to install rst2pdf specially.

**Comment by jaylett** [comment]
Sphinx 1.4 is now out, so I really don't think we can require a 1.2.x series.

I have LaTeX installed to build Xapian core docs in all flavours, and as ojwb says we can provide downloads for everyone else. (Indeed we already do, via RTD, although we want to move to our own hosting at some point so we can build all the language variants.)

**Comment by nirmalsinghania2008** [comment]
@jaylett @ojwb i'll try to get it working with pdflatex soon.

**Comment by ojwb** [comment]
I'm going to close this PR - the current state of rst2pdf sadly makes in unsuitable, and a patch to sort out building with LaTeX would likely have nothing in common with this, so would be best handled as a fresh PR.

But thanks for your work on this @nirmalsinghania2008.

**Pull request closed without merging**